### PR TITLE
TTI: Fix special casing vectorization costs of saturating add/sub

### PR DESCRIPTION
**Pull request by arsenm**

### CODE REVIEW

**Comment by arsenm** [comment]
* **#97463** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/97463?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @arsenm and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Review by sparker-arm — COMMENTED** [review]
I think the BasicTTI changes make sense. But for any backends, such as AArch64, which have already made an attempt to cost the saturating intrinsics, it looks like we're now losing out.

I'm wondering if using `getArithmeticInstrCost` would help in the general sense even though, from a brief look, the AArch64 backend wouldn't handle it properly there either.

**Comment by arsenm** [comment]
> or any backends, such as AArch64, which have already made an attempt to cost the saturating intrinsics, it looks like we're now losing out.

How is this attempting to cost these? This was just adding up the cost of the assumed expansion. I don't think any of this code here adding up a hypothetical expansions is reasonable to have as a default 

> 
> I'm wondering if using `getArithmeticInstrCost` would help in the general sense even though, from a brief look, the AArch64 backend wouldn't handle it properly there either.

I have no idea what's going on with the soup of TTI cost functions. It's like there are 3-10 competing APIs intermixed

**Comment by sparker-arm** [comment]
> I have no idea what's going on with the soup of TTI cost functions.

Fair! The glories of mangling IR instructions, intrinsic calls and ISD nodes all in one, multi-layered, mess :)

> For more precise costs the target can override getIntrinsicInstrCost.

I'm not sure I follow why intrinsics would/should be more precise? In this case, doesn't the intrinsic map directly to the ISD node?

**Comment by davemgreen** [comment]
> I'm not sure I follow why intrinsics would/should be more precise? In this case, doesn't the intrinsic map directly to the ISD node?

It can if it is legal (say a v4i32 sadd_sat under aarch64), but if it is expand (say a i32 sadd_sat under AArch64) the cost model will just say return SingleCallCost. It scalarizes vectors, and if the node is custom lowered it will just return "LT.first * 2".
https://github.com/llvm/llvm-project/blob/298e292a76289d93f7c1a80c26e354830c6080e4/llvm/include/llvm/CodeGen/BasicTTIImpl.h#L2330

So that's not very accurate, and having a fallback to at least some expansion would seem better for any operation that isn't transformed into a libcall. If an individual target needed something more precise for what it will do, that can be overridden in the target cost model.

**Comment by RKSimon** [comment]
Can we keep the existing cost estimate code, but instead of returning Cost just stash it as a optional<ExpansionCost> value and fallback to it if the ISD legality fails?

**Comment by arsenm** [comment]
> > For more precise costs the target can override getIntrinsicInstrCost.
> 
> I'm not sure I follow why intrinsics would/should be more precise? In this case, doesn't the intrinsic map directly to the ISD node?

Because to compute an exact cost you need to know the exact expansion is going to use. The current code assumes a specific expansion, adds up the costs of those pieces. This overestimate the cost when the underlying operation is actually legal. The default cost for an arbitrary expand ISD opcode is some fixed constant

**Comment by arsenm** [comment]
> Can we keep the existing cost estimate code, but instead of returning Cost just stash it as a optional value and fallback to it if the ISD legality fails?

I'm not sure what it means for the ISD legality to fail?

**Review by davemgreen — COMMENTED** [review]
The `llvm.sadd.sat.i64` Arm costs are a bit odd - I assume because `getTypeLegalizationCost(i64)` will be `{2, i32}`, and it uses the `LT.first * 2` as the `i32` is legal. The other results seem OK but that one doesn't sound right for this operation. I could add an override in the Arm backend if you think that is the best way forward - the other results look better, including the SVE versions that we don't currently have tests for.

**Comment by arsenm** [comment]
> The `llvm.sadd.sat.i64` Arm costs are a bit odd - I assume because `getTypeLegalizationCost(i64)` will be `{2, i32}`, and it uses the `LT.first * 2` as the `i32` is legal. The other results seem OK but that one doesn't sound right for this operation. I could add an override in the Arm backend if you think that is the best way forward - the other results look better, including the SVE versions that we don't currently have tests for.

What do you mean by odd? They went down, but you seem to be saying they are too large?

**Comment by davemgreen** [comment]
I mean the old cost used to be 10, it is now 4, but something closer to 10 would be better. https://godbolt.org/z/8zs7EbzGj.

It would be nice if there was a better default than trying to cost a `llvm.sadd.sat.i64` as `2*2*cost(llvm.sadd.sat.i32)`, as that is not how these operations work.

File: llvm/test/Analysis/CostModel/X86/arith-ssat.ll [hunk 1]
```diff
@@ -73,9 +73,9 @@ define i32 @add(i32 %arg) {
 ;
 ; SSE42-LABEL: 'add'
 ; SSE42-NEXT:  Cost Model: Found an estimated cost of 4 for instruction: %I64 = call i64 @llvm.sadd.sat.i64(i64 undef, i64 undef)
-; SSE42-NEXT:  Cost Model: Found an estimated cost of 9 for instruction: %V2I64 = call <2 x i64> @llvm.sadd.sat.v2i64(<2 x i64> undef, <2 x i64> undef)
```
Comment (RKSimon) [inline llvm/test/Analysis/CostModel/X86/arith-ssat.ll:76]: These are CUSTOM but without cost table entries, so with this patch we'll default to 2 * cost - I'll add them to the x86 cost tables

**Comment by davemgreen** [comment]
How do we want to handle for things like `@llvm.sadd.sat.i64` (on Arm, or i128 on 64bit systems)? With this patch they will be costed as `2 * cost(i32)`. I don't mind adding custom cost model rules if needed, but it feels like something that the generic cost model should be getting more correct.

**Comment by arsenm** [comment]
> How do we want to handle for things like `@llvm.sadd.sat.i64` (on Arm, or i128 on 64bit systems)? With this patch they will be costed as `2 * cost(i32)`. I don't mind adding custom cost model rules if needed, but it feels like something that the generic cost model should be getting more correct.

It's probably not the best default for custom lowering, but once something is custom, backend intervention is kind of unavoidable. If we want to change the default custom cost heuristic, that would be a bigger and separate change.

**Comment by davemgreen** [comment]
I think I agree for custom, but AFAIU this isn't about that exactly (I should probably have picked a better word). It's about `getTypeLegalizationCost(i64) == {i32, 2}`, and whether it is sensible to use an i32 cost for a i64 operation.

We would hit the same problem on an i64 system if `getTypeLegalizationCost(i128) == {i64, 2}`. The same is true for any other larger-than-legal type and I'm not sure if the best answer is to try and get ever target to add the correct costs for every type that isn't legal. It's OK to override custom costs for custom operations but for these it feels like there should be a more generic way to handle them.

I haven't tried it with this patch (and it probably deserves to be a separate revision), but something like this might work, even if it is still not perfect.
```
    const TargetLoweringBase *TLI = getTLI();
    std::pair<InstructionCost, MVT> LT = getTypeLegalizationCost(RetTy);

    if (LT.second.getScalarSizeInBits() >= RetTy->getScalarSizeInBits() &&
        TLI->isOperationLegalOrPromote(ISD, LT.second)) {
      if (IID == Intrinsic::fabs && LT.second.isFloatingPoint() &&
          TLI->isFAbsFree(LT.second)) {
        return 0;
      }

      // The operation is legal. Assume it costs 1.
      return LT.first;
    }

    // If we can't lower fmuladd into an FMA estimate the cost as a floating
    // point mul followed by an add.
    if (IID == Intrinsic::fmuladd)
      return thisT()->getArithmeticInstrCost(BinaryOperator::FMul, RetTy,
                                             CostKind) +
             thisT()->getArithmeticInstrCost(BinaryOperator::FAdd, RetTy,
                                             CostKind);
    if (IID == Intrinsic::experimental_constrained_fmuladd) {
      IntrinsicCostAttributes FMulAttrs(
        Intrinsic::experimental_constrained_fmul, RetTy, Tys);
      IntrinsicCostAttributes FAddAttrs(
        Intrinsic::experimental_constrained_fadd, RetTy, Tys);
      return thisT()->getIntrinsicInstrCost(FMulAttrs, CostKind) +
             thisT()->getIntrinsicInstrCost(FAddAttrs, CostKind);
    }

    // More of these..

    if (!TLI->isOperationExpand(ISD, LT.second)) {
      // If the operation is custom lowered or .. then assume
      // that the code is twice as expensive.
      return (LT.first * 2);
    }

    // Else, assume that we...
```

**Comment by arsenm** [comment]
> We would hit the same problem on an i64 system if `getTypeLegalizationCost(i128) == {i64, 2}`. The same is true for any other larger-than-legal type and I'm not sure if the best answer is to try and get ever target to add the correct costs for every type that isn't legal. It's OK to override custom costs for custom operations but for these it feels like there should be a more generic way to handle them.

So it's a question of what the default custom lowered cost is in the presence of type legalization. The code here is also ignoring the possibility of targets custom lowering with illegal types. It's pretty terrible that this needs to guess at the all the steps legalization might take later

**Comment by davemgreen** [comment]
> So it's a question of what the default custom lowered cost is in the presence of type legalization. The code here is also ignoring the possibility of targets custom lowering with illegal types. It's pretty terrible that this needs to guess at the all the steps legalization might take later

That's kind of the point of the cost-model. I've updated some of these costs in #100988.

IMO it is easy for a target to get the legal operations correct as when people implement support for them they are already thinking about them. Custom should be OK too for the most part. But for expanded nodes the target might not have even thought about them, and the surface area is so much larger. The generic cost model should give a decent cost if it can. Obviously it is better for it to give decent costs for as many cases as it can.

I believe if #100988 goes in then what remains here should LGTM.

**Review by RKSimon — APPROVED** [review]
I've no objections to this if the ARM guys are OK with it - we're setting a precedent that custom costs should be set in the TTI,  so we should probably make that clear with an extra comment in the "!TLI->isOperationExpand(ISD, LT.second)" description and that "cost * 2" is just a fallback. Once we're agreed on this on the other intrinsic costs PRs should be straightforward. I'm also going to make a bit more of an effort to avoid unnecessary custom lowerings and use the TargetLowering generic expansions more.

**Comment by arsenm** [comment]
> I've no objections to this if the ARM guys are OK with it - we're setting a precedent that custom costs should be set in the TTI

I don't see how this is setting a precedent. This is normalizing what happened for most opcodes, other than this set that didn't map the intrinsic to the ISD opcode. Conceptually custom would always require custom handling for an accurate result

**Review by davemgreen — APPROVED** [review]
I think this unfortunately also sets the precedent that some nodes that are _not_ legal/custom also need to be handled by the target.

Having said that the Arm/AArch64 costs look OK to me. LGTM

**Comment by arsenm** [comment]
It doesn't really make sense to me to assume custom is cheaper than expand. I would probably just change the default custom cost to be the same

**Comment by RKSimon** [comment]
in which case why not just drop the `!TLI->isOperationExpand(ISD, LT.second)` case entirely?

**Comment by arsenm** [comment]
> in which case why not just drop the `!TLI->isOperationExpand(ISD, LT.second)` case entirely?

That's a broader scoped change than this patch, which is just trying to send the intrinsic down the normal ISD path.

**Review by RKSimon — APPROVED** [review]
OK - cheers

**Pull request merged**